specs/_features/eip7594/fork-choice.md

```diff
@@ -0,0 +1,191 @@
+```python
+def on_block(store: Store, signed_block: SignedBeaconBlock) -> None:
+    """
+    Run ``on_block`` upon receiving a new block.
+    """
+    block = signed_block.message
+    # Parent block must be known
+    assert block.parent_root in store.block_states
+    # Make a copy of the state to avoid mutability issues
+    state = copy(store.block_states[block.parent_root])
+    # Blocks cannot be in the future. If they are, their consideration must be delayed until they are in the past.
+    assert get_current_slot(store) >= block.slot
+
+    # Check that block is later than the finalized epoch slot (optimization to reduce calls to get_ancestor)
+    finalized_slot = compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)
+    assert block.slot > finalized_slot
+    # Check block is a descendant of the finalized block at the checkpoint finalized slot
+    finalized_checkpoint_block = get_checkpoint_block(
+        store,
+        block.parent_root,
+        store.finalized_checkpoint.epoch,
+    )
+    assert store.finalized_checkpoint.root == finalized_checkpoint_block
+
+    # Check the block is valid and compute the post-state
+    block_root = hash_tree_root(block)
+    state_transition(state, signed_block, True)
+
+    # [New in EIP7594] Do not import the block if its unrealized justified checkpoint is not available
+    pulled_up_state = state.copy()
+    process_justification_and_finalization(pulled_up_state)
+    assert is_chain_available(store, pulled_up_state.current_justified_checkpoint.root)
+
+    # Add new block to the store
+    store.blocks[block_root] = block
+    # Add new state for this block to the store
+    store.block_states[block_root] = state
+
+    # Add block timeliness to the store
+    time_into_slot = (store.time - store.genesis_time) % SECONDS_PER_SLOT
+    is_before_attesting_interval = time_into_slot < SECONDS_PER_SLOT // INTERVALS_PER_SLOT
+    is_timely = get_current_slot(store) == block.slot and is_before_attesting_interval
+    store.block_timeliness[hash_tree_root(block)] = is_timely
+
+    # Add proposer score boost if the block is timely and not conflicting with an existing block
+    is_first_block = store.proposer_boost_root == Root()
+    if is_timely and is_first_block:
+        store.proposer_boost_root = hash_tree_root(block)
+
+    # Update checkpoints in store if necessary
+    update_checkpoints(store, state.current_justified_checkpoint, state.finalized_checkpoint)
+
+    # Eagerly compute unrealized justification and finality.
+    compute_pulled_up_tip(store, pulled_up_state, block_root)
+```
+
+#### Pull-up tip helpers
+
+##### Modified `compute_pulled_up_tip`
+
+*Note*: Modified to take `pulled_up_state`, the block's state after applying `processing_justification_and_finalization`.
+The application of `processing_justification_and_finalization` now happens in `on_block`.
+
+```python
+def compute_pulled_up_tip(store: Store, pulled_up_state: BeaconState, block_root: Root) -> None:
```

Comment:
Has this function been modified only to avoid executing `process_justification_and_finalization` twice (as it is already executed at line 146 now) or is there any other reason?
If there is no other reason, I think that it is better not to modify this function.